### DISCUSSION TITLE
Check first device class addition

### DISCUSSION THREAD

**Discussion by derjoerg** [kingsleyadam/local-abbfreeathome#31] [Q&A] [answered]
Hi @kingsleyadam,

when you have time, can you check the following commit in my fork https://github.com/derjoerg/local_abbfreeathome/commit/42c36ae1d66815264334e76abb40bb3cda382b5b before I create a PR out of that?

I implement the trigger function, which doesn't has a state, just the function "press()".

What I find a bit awkward are the modifications I needed to make in the freeathome.py file
- so the function _load_switches() - first of all - has a wrong naming (_load_switchActuators), but my new function "_load_triggers()" is nearly a copy of that function
- the function load_devices() always needs to be extended, which also doesn't looks clean

Do you have an idea how to streamline this?
Perhaps just use the "load-devices()" function, create a list at the top of the function with all classes, which can be loaded, iterate over this list and do the "get_devices_by_function" and the create stuff.
The list would now include ["switch_actuator", "trigger"], with this it should be possible to automatically get the correct functionId and also the correct device class.

The additional thing what can be thought about then is a performance improvement. modify "get_devices_by_function()" to "get_devices_by_functions()" and change the single "function_id" also to a list of function_ids, so if we pass several function_ids into this function it only iterates one time over all devices and channels. The _devices return-value can be extended to a dictionary with the function_id as key and all specific devices below.

Just thoughts 😄

**Comment by kingsleyadam** [comment]
Nice, I'll take a look. I did want to rethink a bit how devices got populated to reduce the amount of repeat code this is going to generate.

I'm not too concerned about performance since we will just be referencing an object in memory and won't be repeatedly fetching from the api. I also want to keep the functions as simple as possible for easier testing. But I think you're onto something, I'll try to test some different scenarios.

**Reply by derjoerg** [reply]
As stated, these are just thoughts I wanted to write down as early in the lifetime of this project as possible

**Reply by kingsleyadam** [reply]
Do you think you could share a device that has the `FID_TRIGGER` function? I don't seem to have anything in my F@H setup.

**Reply by derjoerg** [reply]
Here it is:
```
{
  "00000000-0000-0000-0000-000000000000": {
    "devices": {
      "ABB28EBC3651": {
        "interface": "TP",
        "deviceId": "B008",
        "displayName": "Sensor/switch actuator, 8/8gang, MDRC",
        "unresponsive": false,
        "unresponsiveCounter": 0,
        "defect": false,
        "channels": {
           "ch0012": {
            "floor": "01",
            "room": "13",
            "displayName": "Garagentor",
            "selectedIcon": "11",
            "functionID": "45",
            "inputs": {
              "idp0001": {
                "pairingID": 2,
                "value": "1"
              },
              "idp0003": {
                "pairingID": 4,
                "value": "0"
              }
            },
            "outputs": {},
            "parameters": {
              "par0032": "10"
            }
          },
		}
      }
    }
  }
}
```
In the F@H Web-GUI I can configure a actuator as "switch actuator" or as "trigger". A trigger has no state, you just always send a "1" to idp0001.

**Reply by kingsleyadam** [reply, marked as answer]
Ok, here's a PR that should simplify the creation of new devices in the FreeAtHome class.

https://github.com/kingsleyadam/local-abbfreeathome/pull/32

Go ahead and open a PR for the trigger device after this is merged. Hopefully it's clear what I'm attempting to do.

**Reply by derjoerg** [reply]
Looks great

**Comment by kingsleyadam** [comment]
Also note, I've just opened a PR to refactoring the enum naming convention, something to keep in mind in a new device class: https://github.com/kingsleyadam/local-abbfreeathome/pull/36

**Reply by derjoerg** [reply]
Sounds reasonable and looks good.

Can you - perhaps once again - explain why you use the decimal numbers in functions, pairings and parameters? Doesn't this just adds additional overhead e.g. to the development?
When I look at the developer documentation of F@H everything is hex.
When I use swagger everything is hex.
If the ENUMs need to be updated everything needs to be converted to decimal.

Sure decimal is more common for us humans, but if the base (F@H-API) is hex, shouldn't your library not also just use hex?

**Reply by kingsleyadam** [reply]
In the local api the pairing id comes through as a integer. The function id's come through as short hand hex.

Because of this I decided to use integer for minimal amount of extra conversion we have to do.

- Pairing ID doesn't get converted, it's just a straight integer comparison.
- Function ID gets converted to `int(value, 16)` and compared to integer.

If I were to use the full hex value I'd have to run conversion on both.

- `Pairing` ENUM hex value converted to integer and to compare against the API value. (Or API value converted to HEX and compared to ENUM)
- `Function` ENUM hex value converted to integer, API value converted to integer `int(value, 16)` and compared. (OR API value converted from short hand hex to full e.g. `1a` converted to `0x001A` and compared)

I just choose the path of least resistance.
